### ISSUE TITLE
iOS clock animations on homescreen widget

### ISSUE THREAD

**Issue by liudhzhyym** [pawello2222/WidgetExamples#1]
I want to add a homescreen widget to my app, just like the iOS system Clock app, the second hand keeps moving every second. I have been searching on Google and Github for a long time and have not found a suitable solution. I have seen that there are already third-party apps that can implement such desktop widgets. 

Do you know how do they do it?

**Comment by mromanuk** [comment]
> I want to add a homescreen widget to my app, just like the iOS system Clock app, the second hand keeps moving every second. I have been searching on Google and Github for a long time and have not found a suitable solution. I have seen that there are already third-party apps that can implement such desktop widgets.
> 
> Do you know how do they do it?

I saw the same, I'm building a Widget App and I would like to have the sub second hand on my clock. Did you find out?

**Comment by pawello2222** [comment]
@liudhzhyym @mromanuk Thanks for the question. 

According to Apple it can't be done. 

[Here](https://developer.apple.com/forums/thread/650293) you can find a comment made by a Systems Engineer from Apple:

 > [...] However, you cannot update views to the time other than the relative date text label. So you cannot replicate the Clock widget for example and have second hands on a clock tick by while the widget is visible. Sorry.

Also, from the [Human Interface Guidelines for Widgets](https://developer.apple.com/design/human-interface-guidelines/ios/system-capabilities/widgets/):

> **Keep your widget up-to-date.** To remain relevant and useful, widgets should periodically refresh their information. Widgets don’t support continuous, real-time updates, and the system may adjust the limits for updates depending on various factors. 

My understanding is that the only way for third-party apps to provide such animations is private API. The system Clock Widget does this after all, so it might actually be possible to replicate this behaviour using some private methods that I'm not aware of.

I didn't really look into it myself yet (will probably do when I have more time) but in the meantime you can see this post for a starter:
- [Is there a way to find actual SwiftUI API Documentation (and not just the developer documentation)?](https://stackoverflow.com/q/56528833/8697793)

**Comment by pawello2222** [comment]
@liudhzhyym @mromanuk 

It looks like the private method you were looking for is `_clockHandRotationEffect`:
```swift
extension SwiftUI.View {
  @available(iOS 14.0, macOS 11, *)
  @available(tvOS, unavailable)
  @available(watchOS, unavailable)
  public func _clockHandRotationEffect(_ period: WidgetKit._ClockHandRotationEffect.Period, in timeZone: Foundation.TimeZone, anchor: SwiftUI.UnitPoint = .center) -> some SwiftUI.View
}
```

If you'd like to investigate it more, this code can be found in the WidgetKit framework module.

As I mentioned in the post above, the whole tutorial can be found here:
- [Is there a way to find actual SwiftUI API Documentation (and not just the developer documentation)?](https://stackoverflow.com/q/56528833/8697793)

And here's the path to the original WidgetKit module:

> /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/WidgetKit.framework/Modules/WidgetKit.swiftmodule

Please bear in mind that this is a **_private_** method and not approved by Apple - it may stop working any time or your app might be rejected (even after initially approved).

**Comment by liudhzhyym** [comment]
@pawello2222 Thanks for your reply.

The `_clockHandRotationEffect` method rotates the view continuously. I see that some widget apps have discontinuous second movement in the form of ticks. I would like to ask how to implement the second hand animation of ticks in the widget?
Below, I recorded two short videos, one I made myself, using _ClockHandRotationEffect implements continuous rotation, with a widget made by someone else and a second hand moving in the form of a tick.

Continuous rotation: https://youtu.be/JOL5IyfCcVM
The second hand moves in the form of a tick:  https://youtu.be/DzT5Ktvn5R8
